### PR TITLE
UrlChangeProviders fixes

### PR DESCRIPTION
**Pull request by Starzu**
* Lazy initialization of the UrlChangeProviders
* Improved handling of fragment change in WindowUrlPathChangeProvider

### CODE REVIEW

File: core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala [hunk 1]
```diff
@@ -5,13 +5,16 @@ import io.udash.core.Url
...
-import org.scalajs.dom.Element
+import org.scalajs.dom.{Element, Location}
 import org.scalajs.dom.raw.HashChangeEvent
 
 import scala.scalajs.js
 
 /** Provides information about current URL. */
 trait UrlChangeProvider {
+  /** Enables all required event listeners. */
+  def initialize(): Unit
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala]: Should I invoke it myself? Is it invoked by the framework? Add more docs please.

File: core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala [hunk 2]
```diff
@@ -31,14 +34,17 @@ object WindowUrlFragmentChangeProvider extends UrlChangeProvider {
 
   private val callbacks: js.Array[Url => Unit] = js.Array()
 
-  window.onhashchange = (_: HashChangeEvent) => {
-    callbacks.foreach(_.apply(currentFragment))
+  override def initialize(): Unit = {
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala]: Nitpick: could be just:
```scala
override def initialize(): Unit = window.onhashchange = _ => callbacks.foreach(_.apply(currentFragment))
```

File: core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala [hunk 3]
```diff
@@ -55,44 +61,50 @@ object WindowUrlPathChangeProvider extends UrlChangeProvider {
...
+
+      val target = findLink(event.target.asInstanceOf[Node])
+      if (target != null) {
+        val href = target.getAttribute("href")
+        val location = window.location
+        val newUrl = new JSUrl(href, location.toString)
+        val (samePath, sameHash, sameOrigin) = {
+          (isSamePath(location, newUrl), isSameHash(location, newUrl), isSameOrigin(location, newUrl))
+        }
+        val ignore = event.button != 0 || event.metaKey || event.ctrlKey || event.shiftKey ||
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala]: Extract to a def + comment why

File: core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala [hunk 4]
```diff
@@ -5,13 +5,16 @@ import io.udash.core.Url
...
 import io.udash.utils.Registration
 import org.scalajs.dom
-import org.scalajs.dom.Element
+import org.scalajs.dom.{Element, Location}
 import org.scalajs.dom.raw.HashChangeEvent
 
 import scala.scalajs.js
 
 /** Provides information about current URL. */
 trait UrlChangeProvider {
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala]: Is there a reason not to test both of these providers?

Comment (Starzu) [reply core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala]: It's hard to write any useful unit tests for this. I'm going to move selenium tests from the guide to the core project in the nearest future - I think that's the way to test `UrlChangeProvider`s.

File: core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala (hunk 2 above)
Comment (Starzu) [reply core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala]: Unfortunately 2.11 does not like it ;)

**Pull request merged**